plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+	}
+	hypervisors, err := hypervisors.ExtractHypervisors(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract hypervisors: %v", err)
+	}
+	for _, hypervisor := range hypervisors {
+		o.hypervisors[hypervisor.ID] = hypervisor
+	}
+	return nil
+}
+
+// gatherFlavors collects flavors from the OpenStack API.
+func (o *OpenStack) gatherFlavors() error {
+	page, err := flavors.ListDetail(o.compute, &flavors.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list flavors: %v", err)
+	}
+	flavors, err := flavors.ExtractFlavors(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract flavors: %v", err)
+	}
+	for _, flavor := range flavors {
+		o.flavors[flavor.ID] = flavor
+	}
+	return nil
+}
+
+// gatherServers collects servers from the OpenStack API.
+func (o *OpenStack) gatherServers() error {
+	page, err := servers.List(o.compute, &servers.ListOpts{AllTenants: true}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list servers: %v", err)
+	}
+	servers, err := servers.ExtractServers(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract servers: %v", err)
+	}
+	for _, server := range servers {
+		o.servers[server.ID] = server
+		if o.ServerDiagnotics && server.Status == "ACTIVE" {
+			diagnostic, err := diagnostics.Get(o.compute, server.ID).Extract()
+			o.diag[server.ID] = diagnostic
+			if err != nil {
+				return fmt.Errorf("unable to get diagnostics for server(%v) %v", server.ID, err)
+			}
+		}
+	}
+	return nil
+}
+
+// gatherVolumes collects volumes from the OpenStack API.
+func (o *OpenStack) gatherVolumes() error {
+	//if !o.services.ContainsService(volumeV2Service) {
+	//	return nil
+	//}
+	page, err := volumes.List(o.volume, &volumes.ListOpts{AllTenants: true}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list volumes: %v", err)
+	}
+	v := []volume{}
+	volumes.ExtractVolumesInto(page, &v)
+	for _, volume := range v {
+		o.volumes[volume.ID] = volume
+	}
+	return nil
+}
+
+// gatherStoragePools collects storage pools from the OpenStack API.
+func (o *OpenStack) gatherStoragePools() error {
+	//if !o.services.ContainsService(volumeV2Service) {
+	//	return nil
+	//}
+	results, err := schedulerstats.List(o.volume, &schedulerstats.ListOpts{Detail: true}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list storage pools: %v", err)
+	}
+	storagePools, err := schedulerstats.ExtractStoragePools(results)
+	if err != nil {
+		return fmt.Errorf("unable to extract storage pools: %v", err)
+	}
+	for _, storagePool := range storagePools {
+		o.storagePools[storagePool.Name] = storagePool
+	}
+	return nil
+}
+
+// accumulateServerDiagnostics accumulates statistics from the compute(nova) service.
+func (o *OpenStack) accumulateServerDiagnostics(acc telegraf.Accumulator) {
+	for server_id, diagnostic := range o.diag {
+		tags := tagMap{
+			"server_id": server_id,
+		}
+		fields := fieldMap{
+			"server_id": server_id,
+		}
```

Comment:
Please decide for the one or the other. It's bad habit to have the same value as tag and field. I would opt for using a `tag` if cardinality permits...